### PR TITLE
Advanced privacy-aware bugsnag to fix #659

### LINKED ISSUE #659: Collect bug-reports at runtime
Suggestion: https://fabric.io/kits/android/crashlytics

**To-Do added by SecUpwN:**
- [x] Implement runtime bugcollector that sends stracktraces and as few personal data as possible
- [ ] Under menu `Debugging` do not directly show the running log, but ask the user to select between:

```
[  ] Automatically send debug stacktraces
     (preferred option for better bugfixing)
[  ] Create offline debugging logfile
     (remove sensitive data before posting)
```

If a user selects the second choice, our app should **not** send _any_ bug reports through the web and just reveal the debugger with moving logs as currently shown. We also need a way to change the choice.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28875802-collect-bug-reports-at-runtime?utm_campaign=plugin&utm_content=tracker%2F565750&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F565750&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by SecUpwN**
This PR for #659 adds an advanced, privacy-aware `bugsnag` notifier fork maintained by @larsgrefer at https://github.com/freefair/advanced-bugsnag-android and ensures that we can collect bug-reports at runtime in the future to hopefully eliminate a large number of existing bugs and prevent future ones.

@larsgrefer, please add what your fork collects and why it is privacy-aware to our users. Thanks!

### CODE REVIEW

**Comment by larsgrefer** [comment]
My fork isn't ready yet. I'm currently waiting for bugsnag/bugsnag-android#88, bugsnag/bugsnag-android#89 and bugsnag/bugsnag-android#90

**Comment by SecUpwN** [comment]
@larsgrefer, thanks for notifying. I'm just adding the basic stuff. Please keep this open until finished.

**Comment by SecUpwN** [comment]
@larsgrefer, I guess below errors are being generated because your fork isn't ready? Will wait then.

```
A problem occurred configuring project ':AIMSICD'.
> Could not resolve all dependencies for configuration ':AIMSICD:_systemDebugCompile'.
   > Could not find any matches for io.freefair:bugsnag-android:+ as no versions of io.freefair:bugsnag-android are available.
     Searched in the following locations:
         https://jcenter.bintray.com/io/freefair/bugsnag-android/maven-metadata.xml
         https://jcenter.bintray.com/io/freefair/bugsnag-android/
         https://jitpack.io/io/freefair/bugsnag-android/maven-metadata.xml
         https://jitpack.io/io/freefair/bugsnag-android/
         file:/usr/local/android-sdk/extras/android/m2repository/io/freefair/bugsnag-android/maven-metadata.xml
         file:/usr/local/android-sdk/extras/android/m2repository/io/freefair/bugsnag-android/
         file:/usr/local/android-sdk/extras/google/m2repository/io/freefair/bugsnag-android/maven-metadata.xml
         file:/usr/local/android-sdk/extras/google/m2repository/io/freefair/bugsnag-android/

     Required by:
         Android-IMSI-Catcher-Detector:AIMSICD:0.1.37-alpha
```

File: AIMSICD/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -26,6 +26,11 @@
     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
 
     <!-- SPECIAL PERMISSIONS TO BE ADDED AFTER THIS LINE. /> -->
+    
+    <!-- Required: Used to deliver privacy-aware Bugsnag crash reports -->
+    <uses-permission android:name="android.permission.INTERNET"/>
+    <!-- Optional to see which Activity was active at the time of a crash -->
+    <uses-permission android:name="android.permission.GET_TASKS"/>
```
Comment (larsgrefer) [inline AIMSICD/src/main/AndroidManifest.xml]: When my fork is ready we don't need to declare the permissions here (they will be merged from the bugsnag library during the build, see bugsnag/bugsnag-android#90)

File: AIMSICD/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -65,6 +70,11 @@
                     android:name="android.support.PARENT_ACTIVITY"
                     android:value=".AIMSICD"/>
         </activity>
+        
+        <activity
+                    android:name="com.bugsnag.android.API_KEY"
+                    android:value="9653f64c8477152cc19eb9efb011397d"/>
```
Comment (larsgrefer) [inline AIMSICD/src/main/AndroidManifest.xml]: Are you sure this one should be public?

Comment (SecUpwN) [inline AIMSICD/src/main/AndroidManifest.xml]: Isn't that what is required to be done [here](https://github.com/freefair/advanced-bugsnag-android#configuring-your-androidmanifest)?

File: AIMSICD/src/main/AndroidManifest.xml (hunk 1 above)
Comment (SecUpwN) [inline AIMSICD/src/main/AndroidManifest.xml]: Ok, fair. Would you recommend me to remove these persmissions again, or keep them?

File: AIMSICD/src/main/AndroidManifest.xml (hunk 2 above)
Comment (larsgrefer) [inline AIMSICD/src/main/AndroidManifest.xml]: We have to keep the key private

Comment (SecUpwN) [inline AIMSICD/src/main/AndroidManifest.xml]: >We have to keep the key private

What to add there instead? If I just shall remove this part, I'll do so ASAP and generate a new one.

Comment (larsgrefer) [inline AIMSICD/src/main/AndroidManifest.xml]: @SecUpwN I'll do that, once my fork is ready

Comment (SecUpwN) [inline AIMSICD/src/main/AndroidManifest.xml]: >I'll do that, once my fork is ready

What exactly? Remove that part or generate a fresh API key?

**Comment by SecUpwN** [comment]
@larsgrefer, please give me a sign here when your fork is ready. Thank you.

**Comment by SecUpwN** [comment]
@larsgrefer, all PRs you waited for have been merged. Are you able to resolve the merge conflicts here?

**Comment by SecUpwN** [comment]
@larsgrefer, I'm not sure why you don't answer anymore, hoping you're not mad at me or anything. Should there be anything annoying that currently keeps you from contributing as much as you did before, please send me an email to clear it up. Do you see a chance to finish this PR so that we can add it in our upcoming release to do a step towards resolving as many bugs as possible from our app?

**Comment by larsgrefer** [comment]
Before we can use our special bugsnag client in AIMSICD, I have to implement the client itself

**Comment by SecUpwN** [comment]
>Before we can use our special bugsnag client in AIMSICD, I have to implement the client itself

Oh, I thought we'd just need to add `compile 'io.freefair:bugsnag-android:+'` to our dependencies?

**Comment by larsgrefer** [comment]
That is the plan, but at the moment there is no difference yet between the original `bugsnag-android` and the `advanced-bugsnag-android` except the name

**Comment by SecUpwN** [comment]
@larsgrefer, without being pushy, could you please continue to work on this so that we can merge it within the next two weeks? I'd like to publicly announce that our app now collects anonymized stacktraces for much better error detection when this is done and make sure users can contribute useful logs to finally cut down on our huge list of Issues. My secret hope is also that more devs will join in then. :)

**Comment by agilob** [comment]
bump.

**Comment by agilob** [comment]
bump?

**Comment by SecUpwN** [comment]
> That is the plan, but at the moment there is no difference yet between the original `bugsnag-android` and the `advanced-bugsnag-android` except the name

@larsgrefer, is there still no difference like you said in https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/pull/714#issuecomment-177239345? How can we help you?

**Comment by larsgrefer** [comment]
Im currently to busy to work on the bugsnag client, but for the meantime I've implemented 4d99950

**Comment by SecUpwN** [comment]
> Im currently to busy to work on the bugsnag client, but for the meantime I've implemented 4d99950

Thanks, @larsgrefer. My secret hope is that if we merge this PR when finished, more developers may help us erasing the many bugs our app has. It would also mean that user might not have to post logfiles.

**Pull request closed without merging**